### PR TITLE
Save session Feature

### PR DESCRIPTION
**Pull request by BryceStevenWilley**
First pass at solving #3. Allows admin users to examine user sessions in more detail to reproduce bugs and
strange behavior.

It stores the information in a DARedis database, and also adds a separate
admin-only interview to directly jump to the session in question

Can also save information on the server even if github isn't set up (though not
optimized for that use yet).

### CODE REVIEW

**Review by nonprofittechy — APPROVED** [review]
Looks good! Couple thoughts that can come as we refine this feature:

- Linking a public "review" to a session ID with personal information should be optional. It looks like here the session id is always sent to the feedback form and it is saved to redis every time?
- We may eventually need to refactor this to scale better, but I'm cool with waiting. Right now we get a few dozen feedback items/year so I'm not worried yet. If we get there, SQLObject provides a relatively easy to use frontend to create and manage data in a native SQL database. And Postgres has good support for paging results.

**Comment by BryceStevenWilley** [comment]
I'll add a config that lets server admins turn it on or off. There was some commented out code that let the user themselves choose to link it, but IMO it's not super informative, partially because it suggests that if you say no, server admins can't access your answers (which isn't true if the interview is multi_user).

Hm. SQLObject tbh seems a [lot more complicated](https://docassemble.org/docs/objects.html#SQLObject), and even though I don't expect us to go above 1000 feedback entries a year, it might be a pain to port those to be SQLObjects eventually, and I don't know how well Redis will scale / when it will need to be ported (I expect Redis will scale pretty well actually, but the pickling might slow things down). SQLObjects also seem to be more in line with other types of ORMs, but we don't need an ORM: we need a write once, read a few times sort of thing. Wrapping that in 2-3 functions that just execute SQL (create the table, add an entry, and get a list or subset of entries) matches how this interview will access the DB info a lot closer, and I've done a ton of similar things in the EFSP server, so I might try for that implementation.

We'll need something similar for the thumbs-up/down feedback thing in the toolkit too probably. I might put this on hold until I get something ready for that feature, and then change this to use that feature as well.

**Comment by nonprofittechy** [comment]
> I'll add a config that lets server admins turn it on or off. There was some commented out code that let the user themselves choose to link it, but IMO it's not super informative, partially because it suggests that if you say no, server admins can't access your answers (which isn't true if the interview is multi_user).

I think the distinction is still worth making, because while an admin can view answers on the server, we can't definitively link them to a user's review of our project. This would link otherwise anonymous-seeming feedback to a specific set of answers.

It may be a small distinction at our current usage numbers, because we can link the two pretty well just based on the time of day the feedback came in, but I think it's still meaningful. It may also be something we can use as our own signal and honor it if the user requests to keep the information private.

I'm not sure yet what I think of the best solution, but it does feel like we can have a tiny step to give users choice here.

> Hm. SQLObject tbh seems a [lot more complicated](https://docassemble.org/docs/objects.html#SQLObject), and even though I don't expect us to go above 1000 feedback entries a year, it might be a pain to port those to be SQLObjects eventually, and I don't know how well Redis will scale / when it will need to be ported (I expect Redis will scale pretty well actually, but the pickling might slow things down). SQLObjects also seem to be more in line with other types of ORMs, but we don't need an ORM: we need a write once, read a few times sort of thing. Wrapping that in 2-3 functions that just execute SQL (create the table, add an entry, and get a list or subset of entries) matches how this interview will access the DB info a lot closer, and I've done a ton of similar things in the EFSP server, so I might try for that implementation.
> 
> We'll need something similar for the thumbs-up/down feedback thing in the toolkit too probably. I might put this on hold until I get something ready for that feature, and then change this to use that feature as well.

I think you're right, just directly using the SQL queries might be better. Just use SQLAlchemy or a similar SQL frontend tool for the features that help prevent SQL injection, but we don't need to map it to an object.

**Comment by BryceStevenWilley** [comment]
This wasn't quite ready to merge yet, should have made that a bit clearer, haha. There's at least one bug, I'll patch that in a PR really quick.

**Comment by nonprofittechy** [comment]
Sorry about that--I'll let you merge these PRs in an order that makes sense to you.

File: docassemble/GithubFeedbackForm/data/questions/feedback.yml
Comment (plocket) [inline docassemble/GithubFeedbackForm/data/questions/feedback.yml:292]: Just some small thoughts on language. You don't have to take the whole thing, it's just here if you want it. I feel that if we want the sentiment of "we will not contact you" for the whole thing, that should be in a different place.

```
**Share my answers with an administrators**[BR]
Your answers can help us try to track down the problem. Only server administrators will be able to see your answers.
```

@CaroRob , any thoughts for language on this page?

Comment (BryceStevenWilley) [reply docassemble/GithubFeedbackForm/data/questions/feedback.yml:292]: Thanks! I had made my own version at https://github.com/SuffolkLITLab/docassemble-GithubFeedbackForm/pull/22/files#diff-513abd93f534244b9f400e7681d5fa5ad0b8093562cba85b9cced5dfde618f23R103-R105, which is 

> If you ran into a bug, you can share your answers with
      an administrator (they will not be made public) to help us
      track down the problem.
      (label was next to field, not above)
      
  But I like parts of this as well. Might combine the two:
  
  > **Share my answers with administrators**[BR]
If you ran into a bug, your answers can help us track down the problem. Your answers will not be made public.

**Pull request merged**
